executor/join.go

```diff
@@ -889,12 +911,44 @@ func (e *NestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err e
 			e.hasMatch = false
 			e.hasNull = false
 
-			for _, col := range e.outerSchema {
-				*col.Data = e.outerRow.GetDatum(col.Index, col.RetType)
-			}
-			err = e.fetchAllInners(ctx)
-			if err != nil {
-				return err
+			if e.canUseCache {
+				var key []byte
+				for _, col := range e.outerSchema {
+					*col.Data = e.outerRow.GetDatum(col.Index, col.RetType)
+					key, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, key, *col.Data)
+					if err != nil {
+						return err
+					}
+				}
+				e.cacheAccessCounter++
+				value, err := e.cache.Get(key)
+				if err != nil {
+					return err
+				}
+				if value != nil {
+					e.innerList = value
+					e.cacheHitCounter++
+				} else {
+					err = e.fetchAllInners(ctx)
+					if err != nil {
+						return err
+					}
+					setSuccess, err := e.cache.Set(key, e.innerList)
+					if err != nil {
+						return err
+					}
+					if setSuccess {
+						e.innerList = e.innerList.Copy()
```

Comment:
1. I do not understand why we use `copy` here instead of create a new one?
2. It seems that the `memtracker` of the new `innerList` attach to no one

Reply:
updated.